### PR TITLE
VACMS-8400: Add office hours widget to service location paragraph.

### LINKED ISSUE #8400: Add field_office_hours to paragraphs_type/service_location (reveal on top task) and update all to use HTML5
## User Story or Problem Statement
We need to move from field_facility_service_hours (tablefield)  to field_office_hours in advance of the data migration so that it can be used for top task pages.

## Acceptance Criteria
- [ ] The field_office_hours is added to paragraphs_type/service_location
    - [ ] 12 hour time with am/pm
    - [ ] Comments field
    - [ ] Time increments by 15 minute increments
- [ ] The both field_facility_service_hours and field_office_hours both visible to admin only (ALWAYS EVERYWHERE)
- [ ] (non-admins) The field_office_hours is only visible to editors on vha_facility_nonclinical_service. and field_facility_service_hours is NOT visible.   All other places will continue to show field_facility_service_hours 
- [ ] Save data that is currently in the existing top task pages into the new field. (there is only about 5 of them) once this lands on prod.
- [ ] All Office hours fields are updated to HTML5

## Implementation steps
<optional>

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [x] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [ ] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [x] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [x] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [ ] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team
Please check the team(s) that will do this work. 

- [ ] `CMS Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide CMS Team ` (leave Sitewide unchecked and check the specific team instead)
  - [x] `⭐️ Content ops`
  - [ ] `⭐️ CMS experience`
  - [ ] `⭐️ Offices`
  - [ ] `⭐️ Product support`
  - [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
closes #8400

## Testing done
Visual / Behat

## Screenshots
As an editor / content admin not on a `VAMC Facility Non-clinical Service` page
image

As an admin not on a `VAMC Facility Non-clinical Service` page
image

As an editor / content admin on a `VAMC Facility Non-clinical Service` page
image

Node view
image

## QA steps
**Admin test**
 - [x] Non clinical service: go to `https://pr8435-cunes7rsm9rifo1ypfj5fnybl6l0trl1.ci.cms.va.gov/node/44181/edit`and visually two hours fieldsets appear in the hours section (one table, the other in office hours html5 time inputs format - display will be funky until we get rid of the old table field)
 - [x] Health service: go to `https://pr8435-cunes7rsm9rifo1ypfj5fnybl6l0trl1.ci.cms.va.gov/node/1017/edit` and visually two hours fieldsets appear in the hours section (one table, the other in office hours html5 time inputs format - display will be funky until we get rid of the old table field)
 
**Content Admin / Editor tests**
 - [x] As the same administrator, go to `user/34/workbench_access` and assign Ryan to Chillicothe (this is where a `VAMC Facility Non-clinical Service` exists / can be used for testing)
 ~- [ ] As Ryan.stubblebine@va.gov, go to `node/1017/edit` and visually verify the `Use the facility hours` field does not show hours table when option is set to `Provide specific hours for this service`~ This is a `VAMC Facility Health Service` and is covered by `VAMC Facility Health Service` bundle test in next step
 - [x] as RS  go to this page https://pr8435-cunes7rsm9rifo1ypfj5fnybl6l0trl1.ci.cms.va.gov/node/1058/edit and validate he can still edit / add hours to the facility health service.
 - [x] Login back in as administrator, go to `user/34/edit` and assign Ryan the `Content admin` role
 - [x] Login again as Ryan.stubblebine@va.gov, go to `/node/44181/edit` and visually verify the `Use the facility hours` field shows a single hours table when option is set to `Provide specific hours for this service`
 - [x] on node view of https://pr8435-cunes7rsm9rifo1ypfj5fnybl6l0trl1.ci.cms.va.gov/chillicothe-health-care/locations/chillicothe-va-medical-center-0  no one can see the output from the tablefield.
 - [x] Add some time data to the new hours field, and save
 - [x] Visually verify the node view looks reasonably formatted.
 - [x] time is in 12 hr format matching Vet center cap

**QA Copied from closed pr https://github.com/department-of-veterans-affairs/va.gov-cms/pull/8430**

1. VAMC Facility
2. VAMC System Billing and Insurance
3. Vet Center
4. Vet Center - CAP
5. Vet Center - Mobile
6. Vet Center - Outstation

As cms admin viewing office hours field for the above types, confirm the following settings for office hours field
- [ ] 12 hour time with am/pm
- [ ] Has comments field enabled with HTML disallowed
- [ ] Time increments by 1 minute increments
- [ ] Hours validation is checked

### CODE REVIEW

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 1]
```diff
@@ -495,6 +497,27 @@ public function removeCollapseButton(array &$form) {
...
+  public function removeFieldOfficeHours(array &$form, FormStateInterface $form_state) {
+    /** @var \Drupal\node\NodeForm $form_object $form_object */
+    $form_object = $form_state->getFormObject();
+    /** @var \Drupal\node\Entity\Node $entity */
+    $entity = $form_object->getEntity();
+    if (!empty($form['#paragraph_type'])
+    && $form['#paragraph_type'] === 'service_location'
+    && !$this->userPermsService->hasAdminRole(TRUE)
+    && $entity->bundle() !== 'vha_facility_nonclinical_service') {
+      unset($form['subform']['field_office_hours']);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:517]: I think this logic is forgetting the part about hiding the tablefield version of the field.

**Review by swirtSJW — CHANGES_REQUESTED** [review]
This is not complete but I found this wich is going to require a logic change.

File: docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php [hunk 2]
```diff
@@ -495,6 +497,32 @@ public function removeCollapseButton(array &$form) {
...
+    }
+    if (!empty($form['#paragraph_type'])
+    && $form['#paragraph_type'] === 'service_location'
+    && !$this->userPermsService->hasAdminRole(TRUE)) {
+      unset($form['subform']['field_facility_service_hours']);
```
[the remark this answers is not in the document]
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/src/EventSubscriber/EntityEventSubscriber.php:523]: :+1: updated

**Review by swirtSJW — DISMISSED** [review]
QA checks out for me.   
FInal issue of time format resolved 
image

Nice work.  (just waiting on tests to complete)

**Comment by swirtSJW** [comment]
I think tests on this fell down due to todays github flakyness

**Comment by ethanteague** [comment]
Updated time pickers to html5 per slack convo https://dsva.slack.com/archives/CT4GZBM8F/p1648062097027019
image
image

**Review by kevwalsh — DISMISSED** [review]
There should be only Comment label per table. Not one per row. 


<details><summary>should look like this</summary>

​
image


</details>

**Comment by ethanteague** [comment]
> 

Do you know who created that table, @kevwalsh? I can't find the magic that is creating it (not in tpl / custom code)

**Review by swirtSJW — CHANGES_REQUESTED** [review]
Somethings not right
Yesterday when we had the html5 only the 15 min increments were showing (which was perfect)
image
Today we have all numbers.... but if you don't choose one of the quarters, it fails validation.
image

File: config/sync/core.entity_view_display.paragraph.service_location.default.yml [hunk 3]
```diff
@@ -114,6 +116,33 @@ content:
     third_party_settings: {  }
     weight: 2
     region: content
+  field_office_hours:
+    type: office_hours_table
+    label: above
+    settings:
+      day_format: long
+      time_format: G
```
Comment (swirtSJW) [inline config/sync/core.entity_view_display.paragraph.service_location.default.yml:124]: should this be 'h' like all the others?

Comment (swirtSJW) [reply config/sync/core.entity_view_display.paragraph.service_location.default.yml:124]: or actually shouldnt they all be 'g '  `12-hour format of an hour without leading zeros`?

Comment (ethanteague) [reply config/sync/core.entity_view_display.paragraph.service_location.default.yml:124]: Yeah, I think `g` is correct.

**Comment by swirtSJW** [comment]
It looks like validation is not working right as it should have prevented this
image
I validated the setting is right, 
image
So I am going to call this an upstream issue  and not block this PR on it.

**Review by swirtSJW — APPROVED** [review]
Checks out.
There are two issues noticed but are not regressions introduced here
1) time validation of start time that is after end time does not happen (most likely an upstream issue) https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8469
2) time display on vamc facility pages is in 24 hr time. (Most likely a view mode setting) https://github.com/department-of-veterans-affairs/va.gov-cms/issues/8468
Due to urgency we are not going to block this PR on either issue.  Will follow up later.

**Comment by swirtSJW** [comment]
Woot!  
image

**Pull request merged**